File: spring-data-requery-java/src/main/java/com/coupang/springframework/data/requery/repository/support/SimpleRequeryRepository.java
package com.coupang.springframework.data.requery.repository.support;

import com.coupang.springframework.data.requery.core.RequeryOperations;
import com.coupang.springframework.data.requery.utils.RequeryUtils;
import io.requery.query.*;
import io.requery.query.element.QueryElement;
import io.requery.query.function.Count;
import io.requery.sql.EntityDataStore;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.NotImplementedException;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.domain.*;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashSet;
import java.util.List;
import java.util.Optional;

/**
 * Default implementation of the {@link org.springframework.data.repository.CrudRepository} interface.
 * This will offer you a more sophisticated interface than the plain {@link EntityDataStore}.
 *
 * @author debop@coupang.com
 * @since 18. 6. 4
 */
@Slf4j
@Repository
@Transactional(readOnly = true)
public class SimpleRequeryRepository<T, ID> implements RequeryRepositoryImplementation<T, ID> {

    @Getter
    private final RequeryOperations operations;
    private final RequeryEntityInformation<T, ID> entityInformation;
    private final Class<T> domainClass;
    private final String domainClassName;

    private CrudMethodMetadata crudMethodMetadata;

    public SimpleRequeryRepository(@NotNull RequeryEntityInformation<T, ID> entityInformation, @NotNull RequeryOperations operations) {
        log.debug("Create SimpleRequeryRepository. domainClass={}", entityInformation.getJavaType());

        this.entityInformation = entityInformation;
        this.domainClass = entityInformation.getJavaType();
        this.domainClassName = domainClass.getSimpleName();

        this.operations = operations;
    }

    @Override
    public void setRepositoryMethodMetadata(CrudMethodMetadata crudMethodMetadata) {
        this.crudMethodMetadata = crudMethodMetadata;
    }

    @Transactional
    @Override
    public <S extends T> S upsert(S entity) {
        return operations.upsert(entity);
    }

    @Transactional
    @Override
    public <S extends T> List<S> upsertAll(Iterable<S> entities) {
        return operations.upsertAll(entities);
    }

    @Transactional
    @Override
    public void deleteInBatch(Iterable<T> entities) {
        operations.deleteAll(entities);
    }

    @Transactional
    @Override
    public int deleteAllInBatch() {
        return operations.delete(domainClass).get().value();
    }

    @Override
    public T getOne(ID id) {
        return operations.findById(domainClass, id);
    }

    @NotNull
    @Override
    public List<T> findAll(@NotNull Sort sort) {

        log.debug("Find all {} with sort, sort={}", domainClassName, sort);

        if (sort.isSorted()) {
            OrderingExpression<?>[] orderingExprs = RequeryUtils.getOrderingExpressions(domainClass, sort);

            if (orderingExprs.length > 0) {
                return operations
                    .select(domainClass)
                    .orderBy(orderingExprs)
                    .get()
                    .toList();
            }
        }

        return operations
            .select(domainClass)
            .get()
            .toList();
    }

    @SuppressWarnings("unchecked")
    @NotNull
    @Override
    public Page<T> findAll(@NotNull Pageable pageable) {

        log.debug("Find all {} with paging, pageable={}", domainClassName, pageable);

        if (pageable.isPaged()) {
            QueryElement<? extends Result<T>> query = (QueryElement<? extends Result<T>>)
                RequeryUtils.applyPageable(domainClass,
                                           (QueryElement<? extends Result<T>>) operations.select(domainClass),
                                           pageable);
            List<T> content = query.get().toList();
            long total = operations.count(domainClass).get().value().longValue();

            return new PageImpl<>(content, pageable, total);
        } else {
            List<T> content = operations
                .select(domainClass)
                .get()
                .toList();
            return new PageImpl<>(content);
        }
    }

    @Transactional
    @NotNull
    @Override
    public <S extends T> S save(@NotNull S entity) {
        return operations.upsert(entity);
    }

    @Transactional
    @NotNull
    @Override
    public <S extends T> List<S> saveAll(@NotNull Iterable<S> entities) {
        return operations.upsertAll(entities);
    }

    @NotNull
    @Override
    public Optional<T> findById(@NotNull ID id) {
        return Optional.ofNullable(operations.findById(domainClass, id));
    }

    @SuppressWarnings("unchecked")
    @Override
    public boolean existsById(@NotNull ID id) {
        NamedExpression<ID> keyExpr = (NamedExpression<ID>) RequeryUtils.getKeyExpression(domainClass);

        Tuple result = operations
            .select(Count.count(domainClass))
            .where(keyExpr.eq(id))
            .get()
            .first();

        return result.<Integer>get(0) > 0;
    }

    @NotNull
    @Override
    public List<T> findAll() {
        return operations.findAll(domainClass);
    }

    @SuppressWarnings("unchecked")
    @NotNull
    @Override
    public List<T> findAllById(@NotNull Iterable<ID> ids) {
        HashSet<ID> idSet = new HashSet<>();
        for (ID id : ids) {
            idSet.add(id);
        }
        NamedExpression<ID> keyExpr = (NamedExpression<ID>) RequeryUtils.getKeyExpression(domainClass);

        return operations
            .select(domainClass)
            .where(keyExpr.in(idSet))
            .get()
            .toList();
    }

    @Override
    public long count() {
        return operations.count(domainClass).get().value().longValue();
    }

    @SuppressWarnings("unchecked")
    @Transactional
    @Override
    public void deleteById(@NotNull ID id) {
        log.debug("Delete {} by id. id={}", domainClassName, id);

        NamedExpression<ID> keyExpr = (NamedExpression<ID>) RequeryUtils.getKeyExpression(domainClass);

        Integer deletedCount = operations
            .delete(domainClass)
            .where(keyExpr.eq(id))
            .get()
            .value();

        log.debug("Deleted count={}", deletedCount);
    }

    @Transactional
    @Override
    public void delete(@NotNull T entity) {
        operations.delete(entity);
    }

    @Transactional
    @Override
    public void deleteAll(@NotNull Iterable<? extends T> entities) {
        operations.deleteAll(entities);
    }

    @Transactional
    @Override
    public void deleteAll() {
        log.debug("Delete All {} ...", domainClassName);

        Integer deletedCount = operations.delete(domainClass).get().value();

        log.debug("Delete All {}, deleted count={}", domainClassName, deletedCount);
    }

    @NotNull
    @Override
    public <S extends T> Optional<S> findOne(@NotNull Example<S> example) {
        throw new NotImplementedException("구현 중");
    }

    @NotNull
    @Override
    public <S extends T> List<S> findAll(@NotNull Example<S> example) {
        throw new NotImplementedException("구현 중");
    }

    @NotNull
    @Override
    public <S extends T> List<S> findAll(@NotNull Example<S> example, @NotNull Sort sort) {
        throw new NotImplementedException("구현 중");
    }


    @NotNull
    @Override
    public <S extends T> Page<S> findAll(@NotNull Example<S> example, @NotNull Pageable pageable) {
        throw new NotImplementedException("구현 중");
    }

    @Override
    public <S extends T> long count(@NotNull Example<S> example) {
        throw new NotImplementedException("구현 중");
    }

    @Override
    public <S extends T> boolean exists(@NotNull Example<S> example) {
        throw new NotImplementedException("구현 중");
    }

    @Override
    public Optional<T> findOne(Return<? extends Result<T>> whereClause) {
        return Optional.ofNullable(whereClause.get().firstOrNull());
    }

    @Override
    public List<T> findAll(Return<? extends Result<T>> whereClause) {
        return whereClause.get().toList();
    }

    @SuppressWarnings("unchecked")
    @Override
    public Page<T> findAll(QueryElement<? extends Result<T>> whereClause, Pageable pageable) {

        int total = count(whereClause);

        Return<?> query = RequeryUtils.applyPageable(domainClass, whereClause, pageable);
        List<T> contents = ((Return<? extends Result<T>>) query).get().toList();

        return new PageImpl<>(contents, pageable, total);
    }

    @SuppressWarnings("unchecked")
    @Override
    public List<T> findAll(Iterable<Condition<T, ?>> conditions, Sort sort) {

        QueryElement<? extends Result<T>> baseQuery = (QueryElement<? extends Result<T>>) getOperations().select(domainClass);
        baseQuery = (QueryElement<? extends Result<T>>) RequeryUtils
            .buildWhereClause(baseQuery,
                              RequeryUtils.getGenericConditions(conditions),
                              true);

        Return<? extends Result<T>> query = (Return<? extends Result<T>>) RequeryUtils.applySort(domainClass, baseQuery, sort);

        return query.get().toList();
    }

    @SuppressWarnings("unchecked")
    @Override
    public List<T> findAll(Iterable<Condition<T, ?>> conditions) {
        QueryElement<?> baseQuery = (QueryElement<?>) getOperations().select(domainClass);


        Return<? extends Result<T>> query = (Return<? extends Result<T>>)
            RequeryUtils.buildWhereClause(baseQuery, RequeryUtils.getGenericConditions(conditions), true);

        return query.get().toList();
    }

    @SuppressWarnings("unchecked")
    @Override
    public int count(QueryElement<? extends Result<T>> whereClause) {
        return getOperations().count(domainClass, whereClause);
    }

    @Override
    public boolean exists(QueryElement<? extends Result<T>> whereClause) {
        return getOperations().exists(domainClass, whereClause);
    }


}

File: src/main/kotlin/com/coupang/springframework/data/requery/repository/support/SimpleRequeryRepository.kt
package com.coupang.springframework.data.requery.repository.support

import com.coupang.kotlinx.logging.KLogging
import com.coupang.springframework.data.requery.core.RequeryOperations
import com.coupang.springframework.data.requery.domain.AbstractPersistable
import com.coupang.springframework.data.requery.util.paging
import com.coupang.springframework.data.requery.util.toRequeryOrder
import io.requery.kotlin.`in`
import io.requery.kotlin.eq
import io.requery.query.Result
import io.requery.query.Selection
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Repository
import java.io.Serializable
import java.util.*

/**
 * The implementation of all CRUD, paging and sorting functionality in ArangoRepository from the Spring Data Commons
 * CRUD repository and PagingAndSorting repository
 *
 * @author debop@coupang.com
 * @since 18. 5. 29
 */
@Repository
class SimpleRequeryRepository<T, ID>(
    private val operations: RequeryOperations,
    private val entityType: Class<T>
): RequeryRepositoryImplementation<T, ID> where T: AbstractPersistable<ID>, ID: Serializable {

    override fun findOne(selection: Selection<Result<T>>): T? {
        return selection.get().firstOrNull()
    }

    override fun findAll(selection: Selection<Result<T>>): List<T> {
        return selection.get().toList()
    }

    override fun findAll(selection: Selection<Result<T>>, pageable: Pageable): Page<T> {
        val contents = selection.paging(entityType, pageable).get().toList()
        val totalCount = count(selection)
        return PageImpl(contents, pageable, totalCount)
    }

    override fun findAll(selection: Selection<Result<T>>, sort: Sort): List<T> {
        return selection.orderBy(*sort.toRequeryOrder(entityType)).get().toList()
    }

    override fun count(selection: Selection<Result<T>>): Long {
        // return selection.get().count().toLong()

        return operations.dataStore
            .count(entityType)
            .where()
            .exists(selection)
            .get()
            .value()
            .toLong()
    }

    companion object: KLogging()

    // private val exampleConverter = RequeryExampleConverter(operations.converter.mappingContext)

    override fun findAll(): Iterable<T> {
        return operations.findAll(entityType)
    }

    override fun findAll(sort: Sort): Iterable<T> {
        return operations.dataStore.select(entityType).orderBy(*sort.toRequeryOrder(entityType)).get().toList()
    }

    //    override fun <S: T> findAll(example: Example<S>): Iterable<S> {
    //        TODO("not implemented")
    //    }
    //
    //    override fun <S: T> findAll(example: Example<S>, sort: Sort): Iterable<S> {
    //        TODO("not implemented")
    //    }

    override fun findAll(pageable: Pageable): Page<T> {
        val contents = operations.dataStore.select(entityType).paging(entityType, pageable).get().toList()
        val totalCount = operations.dataStore.count(entityType).get().value().toLong()
        return PageImpl(contents, pageable, totalCount)
    }

    //    override fun <S: T> findAll(example: Example<S>, pageable: Pageable): Page<S> {
    //        TODO("not implemented")
    //    }

    override fun findAllById(ids: Iterable<ID>): Iterable<T> {
        return operations.dataStore
            .select(entityType)
            .where(AbstractPersistable<ID>::id.`in`(ids.toList()))
            .get()
            .toList()
    }

    override fun <S: T> saveAll(entities: Iterable<S>): Iterable<S> {
        return operations.saveAll(entities)
    }

    override fun <S: T> upsert(entity: S): S {
        return operations.upsert(entity)
    }

    override fun <S: T> upsertAll(entities: Iterable<S>): Iterable<S> {
        return operations.upsertAll(entities)
    }

    override fun deleteInBatch(entities: Iterable<T>) {
        operations.deleteAll(entities)
    }

    override fun deleteAllInBatch() {
        operations.deleteAll(entityType)
    }

    override fun getOne(id: ID): T? {
        return operations.findById(entityType, id)
    }

    override fun <S: T> save(entity: S): S {
        return operations.save(entity)
    }

    override fun deleteById(id: ID) {
        operations.dataStore
            .delete(entityType)
            .where(AbstractPersistable<ID>::id.eq(id))
            .get()
    }

    override fun deleteAll(entities: MutableIterable<T>) {
        operations.deleteAll(entities)
    }

    override fun deleteAll() {
        operations.deleteAll(entityType)
    }

    override fun count(): Long {
        return operations.count(entityType).get().value().toLong()
    }

    //    override fun <S: T> count(example: Example<S>): Long {
    //        TODO("not implemented")
    //    }

    override fun existsById(id: ID): Boolean {
        return operations.dataStore
            .count(entityType)
            .where(AbstractPersistable<ID>::id.eq(id))
            .get()
            .value() == 1
    }

    override fun findById(id: ID): Optional<T> {
        return Optional.ofNullable(operations.findById(entityType, id))
    }

    override fun delete(entity: T) {
        operations.delete(entity)
    }

    //    override fun <S: T> findOne(example: Example<S>): Optional<S> {
    //        TODO("not implemented")
    //    }
    //
    //    override fun <S: T> exists(example: Example<S>): Boolean {
    //        TODO("not implemented")
    //    }
}